`devShellTools`: add environment functions

roberth opened the pull request:
## Description of changes

Moves the environment functions, adds tests, docs, and fixes.
See commit messages.

- Builds on https://github.com/NixOS/nixpkgs/pull/323398

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
devShellTools.{unstructuredDerivationInputEnv,derivationOutputEnv}: extract

roberth committed:
devShellTools.unstructuredDerivationInputEnv: Return attrsOf str

... and test it.

roberth committed:
devShellTools.unstructuredDerivationInputEnv: Skip args

roberth committed:
devShellTools: Docs, fix args env

roberth committed:
devShellTools.unstructuredDerivationInputEnv: Match passAsFile basename

roberth committed:
nixosTests.docker-tools-nix-shell: Extract

These tests should run as part of the `tests.devShellTools` suite,
and they're a significant portion of the general docker tools test
duration.

roberth committed:
build-support/docker/default.nix: Refer to docs and tests

This removes redundant inline docs, because
- users should consult the better docs in the manual,
- contributors should add to the manual, not the inline comments

roberth commented:
@ofborg build tests.devShellTools

Included in the above, but check it separately this time:

@ofborg build nixosTests.docker-tools-nix-shell

roberth committed:
devShellTools.unstructuredDerivationInputEnv: Support older Nix

roberth committed:
tests.devShellTools: Fix release.nix / ofborg eval

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4215

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/little-bites-of-nix-with-mise-seeking-design-feedback/48785/2

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-07-08-nix-team-meeting-minutes/49099/1

tie reviewed (commented):
Tested with `nix build --file . tests.devShellTools` and also `nix build --file . nixosTests.docker-tools-nix-shell` on `aarch64-linux` after adjusting it to build for all systems.

roberth committed:
nixosTests.docker-tools-nix-shell: Enable on aarch64-linux

The docker-tools test, where this originates, was not run on aarch64-linux, but this is an artifact of its age more so than anything else.

Co-authored-by: Ivan Trubach <mr.trubach@icloud.com>

roberth committed:
Format

mightyiam commented:
Thank you, @roberth .

roberth commented:
- Next steps: https://github.com/NixOS/nixpkgs/pull/330822
